### PR TITLE
gr-filter/fft_filter: Use smart instead of raw pointers

### PR DESCRIPTION
**Pull request by ThomasHabets**
This reduces amount of code, and risk of dangling pointers and memory leaks.

I thought I'd do this one to get opinions. It takes a while (especially with custom deleter), so don't expect me to do all of it at least at this point. :-)

### CODE REVIEW

**Comment by hcab14** [comment]
Looks good. Just to let you know that there is a recent addition to volk
called volk::vector which helps to simplify the fft code.

**Comment by ThomasHabets** [comment]
Thanks for the pointer. Looks like I'll have to wait for gnuradio to pick up https://github.com/gnuradio/volk/commit/4cc47ea71b4c7d3f9433156f5f61b6c7e11b4f53 though.

**Comment by ThomasHabets** [comment]
Ack. I'm holding off on this until https://github.com/gnuradio/gnuradio/pull/2981 is merged.

**Comment by marcusmueller** [comment]
I mean, this is seriously an improvement. However, I'm still not convinced the fftw plan has to be handled via pointer at all – it's not like anything else would be using it. 

@ThomasHabets I'm certain you're more intimate with the details of the FFT handling by now, but what's the advantage of having a unique pointer instead of holding the raw object?

**Comment by mbr0wn** [comment]
Going pointer -> smart pointer is something I would consider refactoring, and going pointer -> no pointer is something I would consider a code change. If we want the latter, the former is no longer necessary. However, the refactoring by itself makes a lot of sense.

File: gr-filter/lib/fft_filter.cc [hunk 1]
```diff
@@ -24,6 +24,7 @@
 #include "config.h"
 #endif
 
+#include <boost/smart_ptr/make_unique.hpp>
```
Comment (mbr0wn) [inline gr-filter/lib/fft_filter.cc]: Meh. I wonder if we should go straight to C++14, or create our own `std::make_unique` for C++11 as a stop-gap.

**Comment by mbr0wn** [comment]
Personally, I think I prefer the pointers though.

**Comment by mbr0wn** [comment]
But please cut the commit subject to 72 characters.

**Comment by ThomasHabets** [comment]
@marcusmueller looks to me like the plans are not copyable (because in their default the member pointers will be copied by value). It's actually another cleanup I've had in mind. It becomes easier with (again) C++11 and move semantics to fix these.

Currently `fft::fft_real_fwd` and friends appear to be a bit of a minefield, since they don't prohibit copy assignment or copy constructor at compile time, but they leave dangling pointers if called at runtime. (unless I'm missing something)

So yeah, implementing move semantics on the ffts, disabling copy assignment and copy constructor, and de-pointering `fft_filter_*` seems like a good next step.

De-pointering requires some further rework of `fft_filter_*` though: The fft size is not known at construction time, so would need to be broken out into its own function so that it can be provided on the constructor initializer list. Not hard. it's a direct function of the tap count.

Any preferences? In the mean time I'll sort out the volk vector thing. Now that the other pull request is in it seems like that one can go in this commit, right?

**Comment by mbr0wn** [comment]
I think it's not worth de-pointerizing this. Your commit seems to take care of @marcusmueller's question and I like this PR.

**Comment by mbr0wn** [comment]
I also don't think the PR failure is your fault.

**Comment by mbr0wn** [comment]
@ThomasHabets do we have a  CLA from you?

**Comment by ThomasHabets** [comment]
@mbr0wn yup, CLA sorted during https://github.com/gnuradio/gnuradio/pull/2893. Google corp CLA.

**Comment by ThomasHabets** [comment]
There. Now using `volk::vector`. Should be ready to go.

File: gr-filter/include/gnuradio/filter/fft_filter.h [hunk 2]
```diff
@@ -174,13 +173,13 @@ class FILTER_API fft_filter_ccc
     int d_ntaps;
     int d_nsamples;
     int d_fftsize; // fftsize = ntaps + nsamples - 1
-    int d_decimation;
-    fft::fft_complex* d_fwdfft;     // forward "plan"
-    fft::fft_complex* d_invfft;     // inverse "plan"
+    const int d_decimation;
```
Comment (marcusmueller) [inline gr-filter/include/gnuradio/filter/fft_filter.h]: We coulda/shoulda `unsigned` this, but for now, come on, let's get this merged.

**Comment by marcusmueller** [comment]
Merging as soon as CI passes :)

**Comment by marcusmueller** [comment]
Thomas, I have yet to express my gratitude: Thanks for all these contributions!

File: gr-filter/include/gnuradio/filter/fft_filter.h [hunk 3]
```diff
@@ -26,6 +26,7 @@
 #include <gnuradio/fft/fft.h>
 #include <gnuradio/filter/api.h>
 #include <gnuradio/gr_complex.h>
+#include <volk/volk_alloc.hh>
```
Comment (marcusmueller) [inline gr-filter/include/gnuradio/filter/fft_filter.h]: How did CI not catch this? There's an `h` to much...

**Comment by marcusmueller** [comment]
@noc0lour  just a heads up: buildbot somehow missed the above typo, which definitely should have killed the build, somehow. I don't understand how that happens?

File: gr-filter/include/gnuradio/filter/fft_filter.h (hunk 3 above)
Comment (ThomasHabets) [reply gr-filter/include/gnuradio/filter/fft_filter.h]: Is it? It's not just the indicator saying it's a C++-only header?

Comment (ThomasHabets) [reply gr-filter/include/gnuradio/filter/fft_filter.h]: Ah, I see you reverted that commit now.

**Pull request merged**